File: hrm-portal-frontend/src/services/roleService.js
import api from "./api";

const roleService = {

  async getRoles({
    search = "",
    page = 0,
    size = 20
  } = {}) {

    const params = {
      page,
      size
    };

    if (search?.trim()) {
      params.search = search.trim();
    }

    const response = await api.get(
      "/roles",
      { params }
    );

    return response.data;
  },


  async getRoleById(roleId) {

    if (!roleId) {
      throw new Error("Role ID is required.");
    }

    const response = await api.get(
      `/roles/${roleId}`
    );

    return response.data;
  },


  async getPermissions() {

    const response = await api.get(
      "/permissions"
    );

    return response.data;
  },


  async createRole(payload) {

    const response = await api.post(
      "/roles",
      payload
    );

    return response.data;
  },


  async updateRole(
    roleId,
    payload
  ) {

    if (!roleId) {
      throw new Error("Role ID is required.");
    }

    const response = await api.put(
      `/roles/${roleId}`,
      payload
    );

    return response.data;
  },


  async updateRolePermissions(
    roleId,
    permissions
  ) {

    if (!roleId) {
      throw new Error("Role ID is required.");
    }

    const response = await api.put(
      `/roles/${roleId}/permissions`,
      {
        permissions
      }
    );

    return response.data;
  },


  async updateRoleStatus(
    roleId,
    active
  ) {

    if (!roleId) {
      throw new Error("Role ID is required.");
    }

    const response = await api.patch(
      `/roles/${roleId}/status`,
      {
        active
      }
    );

    return response.data;
  },


  async deleteRole(roleId) {

    if (!roleId) {
      throw new Error("Role ID is required.");
    }

    const response = await api.delete(
      `/roles/${roleId}`
    );

    return response.data;
  }

};


export default roleService;

File: src/main/java/com/workforce/hrm/service/RoleService.java
package com.workforce.hrm.service;

import java.util.List;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import com.workforce.hrm.dto.request.RoleRequestDTO;
import com.workforce.hrm.dto.response.PermissionResponseDTO;
import com.workforce.hrm.dto.response.RoleResponseDTO;

public interface RoleService {

    Page<RoleResponseDTO> getRoles(String search, Pageable pageable);

    RoleResponseDTO getRoleById(Long roleId);

    RoleResponseDTO createRole(RoleRequestDTO request);

    RoleResponseDTO updateRole(Long roleId, RoleRequestDTO request);

    RoleResponseDTO updateRolePermissions(Long roleId, List<String> permissions);

    RoleResponseDTO updateRoleStatus(Long roleId, Boolean active);

    void deleteRole(Long roleId);

    List<PermissionResponseDTO> getAllPermissions();
}
